### PR TITLE
Add support for configurable notification comments

### LINKED ISSUE #5: Add support for configurable notifications
Github doesn't natively give granular level of control for notifications on subdirectories. If an active contributor wants to get notified of open PR or issues that only effect a subset of the repo their only option right now is to subscribe to the whole project and get emails for every piece of activity. Using codeowners has been tried to solve this problem for those users who have write access to the repo, but it's not ideal because this couples conditional notifications with approval authority.

Adding a feature to qiskit-bot should not be hard to implement this, it already has all the capabilities to implement it. When qiskit-bot is configured to watch a repo it already knows how to read a local config file (for configurable tags) and it also gets events for every event. The missing piece is the step to take in PR and issue events and check them against the local_config_file for a configured element to see if a user should be notified. If there are users who need to be notified the bot can leave a mention for each configured user as a comment on the PR or issue. This will then generate a notification email for users with github's default notification settings.

### PR DESCRIPTION
**Pull request by mtreinish**
This commit adds a new feature for enabling per repo configurable
notification comments to qiskit-bot. When a repository has a per repo
configuration file (ie qiskit_bot.yaml in the repo root) with the
notifications section set this can be used to provide a list of
usernames that will get mentioned in a generated comment automatically
based on the files modified in the PR. This lets each project have a
local configuration of who gets notified of changes to each submodule in
the repository.

Fixes #5

### CODE REVIEW

**Comment by mtreinish** [comment]
Two things I just realized I should add to this before we merge. First I should add a section to the readme on the per repo `qiskit_bot.yaml` config file. Right now it's not clear how this can get used by projects to opt-in to notifications (or to configure the changelog labels which was a pre-existing feature). The other thing is I should add an option to have the bot always leave a comment, if there isn't a configured user for notification it can just print the prelude. This is useful for having initial communication to the user.

**Comment by mtreinish** [comment]
For an example of this in practice see: https://github.com/mtreinish/qiskit-terra-sandbox/pull/15#issuecomment-1031945765 where I tested this with the bot in a local deployment.

File: qiskit_bot/notifications.py
Comment (manoelmarques) [inline qiskit_bot/notifications.py]: Before your PR can be merged it will first need to run, pass continuous

File: qiskit_bot/notifications.py
Comment (manoelmarques) [inline qiskit_bot/notifications.py]: integration tests and also be reviewed. Sometimes the review process can

Comment (mtreinish) [reply qiskit_bot/notifications.py]:
```suggestion
integration tests and also be reviewed. Sometimes the review process can
```

File: qiskit_bot/notifications.py
Comment (mtreinish) [reply qiskit_bot/notifications.py]: I'm not sure I understand the comma here. What about

```suggestion
Before your PR can be merged it will first need to pass continuous
```

the "run" here was redundant anyway.

File: README.md
Comment (jakelishman) [inline README.md]:
```suggestion
The other main feature the bot offers is to automatically leave a message on
```

File: README.md
Comment (jakelishman) [inline README.md]:
```suggestion
contents. Currently this configuration file is used to control two things:
the changelog generation behavior, and whether the bot will leave comments
on new pull requests when they're opened (and the exact behavior of that
comment). An example of a fully populated configuration file is:
```

File: README.md
Comment (jakelishman) [inline README.md]: This might want updating with other typo fixes.

File: qiskit_bot/notifications.py
Comment (jakelishman) [inline qiskit_bot/notifications.py]: How much do we care about this stuff being accurate?  I think Nature have a checker for the copyrights, which might be nice to use for consistency across Qiskit in the future.

File: qiskit_bot/notifications.py
Comment (jakelishman) [inline qiskit_bot/notifications.py]: Is this docstring accurate?

File: qiskit_bot/notifications.py
Comment (jakelishman) [reply qiskit_bot/notifications.py]: Your new form looks better to me. I think before it read like it was a list of three things, but the Oxford comma might have been more correct in that situation.

At any rate, the word "also" in "also be reviewed" is probably unnecessary.

File: qiskit_bot/notifications.py
Comment (jakelishman) [inline qiskit_bot/notifications.py]:
```suggestion
While you're waiting, please feel free to review other open
PRs. While only a subset of people are authorized to approve pull requests for
merging, everyone is encouraged to review open pull requests. Doing reviews
helps reduce the burden on the core team and helps make the project's code
better for everyone.
```

Comment (jakelishman) [reply qiskit_bot/notifications.py]: (I doubt many people are going to read it, but probably best be as concise as possible to improve our odds.)

File: qiskit_bot/notifications.py
Comment (jakelishman) [inline qiskit_bot/notifications.py]: It's probably going to be nicer for users if we reflow this into continuous text, rather than dumping out a hard-wrapped block.

File: README.md
Comment (jakelishman) [inline README.md:117]: It might be better to use/mention single quotes for regex strings - my understanding is that single-quoted YAML strings don't parse escape characters, but double-quotes do, and literal `\` is needed in regexes fairly commonly.

**Review by jakelishman — COMMENTED** [review]
Are we planning to use this sort of user notification on Terra/other parts of Qiskit?  What's the advantage of this user notification over the current `CODEOWNERS`?

**Comment by mtreinish** [comment]
> Are we planning to use this sort of user notification on Terra/other parts of Qiskit? What's the advantage of this user notification over the current `CODEOWNERS`?

Yeah, that's the basic idea. The biggest advantage is that it decouples notification for a review request from merge permission. A codeowner entry always gets you both, but there might be people that want to explicitly get pinged for PRs for some subsection of the repo but don't have write permission yet. Another category is for splitting up notification of a core team, if the core team has merge permission on the repo but would only like to actively be pinged as an individual on a subtree this can be leveraged for that. To a smaller degree the advantage this provides is the regex match is more flexible than the selection the codeowners file does but I'm not sure how many people are going to want to leverage that (personally I'm looking forward to being notified on every file that has two `'a'`s in the path :P)

File: qiskit_bot/notifications.py
Comment (mtreinish) [reply qiskit_bot/notifications.py]: Nope, it was sloppy copy and paste. It should be something like:
```suggestion
def trigger_notifications(pr_number, repo, conf):
    """Process any potential notifications on a new PR."""
```

**Comment by jakelishman** [comment]
Yeah, that sounds like solid reasoning to me.  I guess it doesn't affect me (or you) personally too much, since we're already subscribed to all PRs made, but I think it'll be a lot of help for other people in the team.

File: qiskit_bot/notifications.py
Comment (mtreinish) [inline qiskit_bot/notifications.py]:
```suggestion
integration tests and be reviewed. Sometimes the review process can
```

File: README.md
Comment (mtreinish) [reply README.md:117]: Yeah, that's a fair point I had forgot that detail from the yaml spec (https://yaml.org/spec/1.2.2/#double-quoted-style vs https://yaml.org/spec/1.2.2/#single-quoted-style for reference). I'll update this to mention that

File: qiskit_bot/notifications.py
Comment (mtreinish) [reply qiskit_bot/notifications.py]: It honestly isn't really enforceable for anything. Also `(C)` is meaningless it's not actually a copyright symbol.  It's just busy work imo to keep updating things on every when we have a complete git log to show modification history. For source distribution without the git history outside of the copyright statement itself the year doesn't really matter too much. That being said I can just fix this easily enough

```suggestion
# (C) Copyright IBM 2022
```

File: README.md
Comment (mtreinish) [reply README.md]: Yeah, I'll update it here and in the tests after the last commit

File: qiskit_bot/notifications.py
Comment (jakelishman) [reply qiskit_bot/notifications.py]: Yeah, on qutip I literally just removed all these notices for that reason (qutip/qutip#1579).  The licence in the repo root is sufficient if it's distributed with the package.

File: README.md
Comment (jakelishman) [inline README.md]:
```suggestion
qiskit-bot will read it before every action and adjust behavior based on its
```

File: README.md
Comment (jakelishman) [reply README.md:117]: Don't know if you want to bother with this (explaining single quotes for YAML regexes).  On a second reading, I'm not sure this README is the right place to include a YAML tutorial.

File: README.md
Comment (jakelishman) [reply README.md]: Just a reminder to update this.

File: README.md
Comment (jakelishman) [inline README.md:120]: Did you also consider having the bot assign the issue instead of a notification comment?  I guess assignment feels more natural, but can't apply to teams.

File: qiskit_bot/config.py
Comment (jakelishman) [inline qiskit_bot/config.py:68]: This doesn't matter, just a question: are there any reasons to define `categories` with a default here, but leave the others default-less?

File: qiskit_bot/notifications.py
Comment (jakelishman) [inline qiskit_bot/notifications.py:42]: Again, not important at all: is there an advantage to defining this here rather than as part of the schema/configuration file?

(I'm guessing testing?)

File: qiskit_bot/notifications.py
Comment (jakelishman) [inline qiskit_bot/notifications.py:69]: I think this probably is irrelevant, because we'll always be running the bot on Linux machines, but is there a potential path-separator issue brewing here on Windows?  If you want to match a subdirectory, I don't know what kind of path names the GitHub interface is going to return - will they be `/`-separated or `\`-separated on Windows?

(I'm also happy just to ignore the comment and say "the bot only runs on unix-likes".)

**Review by jakelishman — COMMENTED** [review]
This all looks fine to me.  Aside from the typo fix, most of the comments are minor questions rather than fixes.

File: README.md
Comment (mtreinish) [reply README.md:120]: I did, but the main issue with this is that assignment is limited to project members only and if anyone wanted to receive a notification outside of being an explicit project team member assignment was not really an option

Comment (jakelishman) [reply README.md:120]: Yep, sounds good.  I'd assumed you'd thought about it, just wondering the reasons.  Now you mention it, I think I knew this reason already sorry.

File: qiskit_bot/notifications.py
Comment (mtreinish) [reply qiskit_bot/notifications.py:69]: At least in production we only run this on linux (specifically debain bullseye using the official python docker image as the base). I'm not sure there is value in trying to support this on windows as I find it extremely unlikely that anyone would want to deploy the bot to a windows environment.

That being said I did think about this a long time ago and made the decision to only list linux as a supported environment in the package metadata trove classifiers: https://github.com/Qiskit/qiskit-bot/blob/master/setup.cfg#L13 although I don't know how many people actually look at those

File: qiskit_bot/notifications.py
Comment (mtreinish) [reply qiskit_bot/notifications.py:42]: Yeah, it's mostly testing I think, but I honestly wrote this long enough ago I'm not really sure

File: qiskit_bot/config.py
Comment (mtreinish) [reply qiskit_bot/config.py:68]: If there is a sane implicit default value not really, just sloppy style on my part mostly. I think I might have had some trouble with voluptuous populating defaults at some point and just opted not to use it if I couldn't get it to work reliably (but that might have just been user error, I don't really remember)

**Review by jakelishman — APPROVED** [review]

**Pull request merged**
